From 45c2d86524c38649ee4cfb83e7e6f52573e6ef67 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 28 Sep 2023 17:38:04 +0200
Subject: [PATCH] board: solidrun: am642-som: runtime detection for eMMC & SD
 environment storage

SolidRun AM642 SoM can boot from either eMMC or SD.
When booting from SD, environment is stored as a file on FAT partition.
When booting from eMMC, environment is stored on boot partition at an
offset.

Implement env_get_location function to select correct environment driver
(FAT / MMC sectors), and mmc_get_env_dev to choose correct mmc device,
based on what media was used to start u-boot..

Additionally enable the required u-boot configuration options.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 .../solidrun/am64x_solidrun/am64x_solidrun.c  | 54 +++++++++++++++++++
 configs/am64x_a53_solidrun_defconfig          |  8 +--
 configs/am64x_r5_solidrun_defconfig           |  8 ++-
 3 files changed, 66 insertions(+), 4 deletions(-)

diff --git a/board/solidrun/am64x_solidrun/am64x_solidrun.c b/board/solidrun/am64x_solidrun/am64x_solidrun.c
index a90004bc34..c2cf2222dd 100644
--- a/board/solidrun/am64x_solidrun/am64x_solidrun.c
+++ b/board/solidrun/am64x_solidrun/am64x_solidrun.c
@@ -11,6 +11,7 @@
 #include <command.h>
 #include <asm/io.h>
 #include <env.h>
+#include <env_internal.h>
 #include <spl.h>
 #include <asm/arch/hardware.h>
 #include <asm/arch/sys_proto.h>
@@ -538,4 +539,57 @@ int find_devicetree(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[
 /*expose functions to uboot shell*/
 U_BOOT_CMD(find_dt, 1, 0, find_devicetree, "", "");
 
+/*
+ * select environment driver by boot source:
+ * - eMMC: raw sector
+ * -   SD: fat partition
+ */
+enum env_location env_get_location(enum env_operation op, int prio)
+{
+	u32 bootdevice = spl_boot_device();
+
+	/* only support prio 0 (one location) */
+	if (prio > 0)
+		return ENVL_UNKNOWN;
+
+	switch(bootdevice) {
+#ifdef CONFIG_ENV_IS_IN_MMC
+		case BOOT_DEVICE_EMMC:
+			/* eMMC (sdhci0): raw sector */
+			return ENVL_MMC;
+#endif
+#ifdef CONFIG_ENV_IS_IN_FAT
+		case BOOT_DEVICE_MMC2:
+			/* SD (sdhci1): data partition */
+			return ENVL_FAT;
+#endif
+	}
+
+	printf("warning: boot device 0x%x does not (yet) support environment storage\n", bootdevice);
+	return ENVL_NOWHERE;
+}
+
+/*
+ * Select environment storage mmc device by boot source
+ * - eMMC: sdhci0
+ * -   SD: sdhci1
+ */
+int mmc_get_env_dev(void) {
+	u32 bootdevice = spl_boot_device();
+	switch(bootdevice) {
+#ifdef CONFIG_ENV_IS_IN_MMC
+		case BOOT_DEVICE_EMMC:
+			/* eMMC (sdhci0): raw sector */
+			return 0;
+#endif
+#ifdef CONFIG_ENV_IS_IN_FAT
+		case BOOT_DEVICE_MMC2:
+			/* SD (sdhci1): data partition */
+			return 1;
+#endif
+	}
+
+	return -ENODEV;
+}
+
 #endif //CONFIG_SPL_BUILD
diff --git a/configs/am64x_a53_solidrun_defconfig b/configs/am64x_a53_solidrun_defconfig
index edbe37e897..5fee7c4276 100644
--- a/configs/am64x_a53_solidrun_defconfig
+++ b/configs/am64x_a53_solidrun_defconfig
@@ -10,7 +10,8 @@ CONFIG_NR_DRAM_BANKS=2
 CONFIG_SOC_K3_AM642=y
 CONFIG_K3_ATF_LOAD_ADDR=0x701c0000
 CONFIG_TARGET_AM642_A53_SOLIDRUN=y
-CONFIG_ENV_SIZE=0x20000
+CONFIG_ENV_SIZE=0x80000
+CONFIG_ENV_OFFSET=0x380000
 CONFIG_SYS_SPI_U_BOOT_OFFS=0x300000
 CONFIG_DM_GPIO=y
 CONFIG_SPL_DM_SPI=y
@@ -76,10 +77,11 @@ CONFIG_SPL_OF_CONTROL=y
 CONFIG_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT_NO_COMPRESSION=y
+CONFIG_ENV_IS_NOWHERE=y
 CONFIG_ENV_IS_IN_FAT=y
+CONFIG_ENV_IS_IN_MMC=y
 CONFIG_ENV_FAT_DEVICE_AND_PART=":1"
-CONFIG_SYS_MMC_ENV_DEV=1
-CONFIG_SPL_ENV_IS_NOWHERE=y
+CONFIG_SYS_MMC_ENV_PART=1
 CONFIG_NET_RANDOM_ETHADDR=y
 CONFIG_DM=y
 CONFIG_SPL_DM=y
diff --git a/configs/am64x_r5_solidrun_defconfig b/configs/am64x_r5_solidrun_defconfig
index b37be6f09a..6a91963b56 100644
--- a/configs/am64x_r5_solidrun_defconfig
+++ b/configs/am64x_r5_solidrun_defconfig
@@ -8,7 +8,8 @@ CONFIG_SYS_MALLOC_F_LEN=0x80000
 CONFIG_NR_DRAM_BANKS=2
 CONFIG_SOC_K3_AM642=y
 CONFIG_TARGET_AM642_R5_SOLIDRUN=y
-CONFIG_ENV_SIZE=0x20000
+CONFIG_ENV_SIZE=0x80000
+CONFIG_ENV_OFFSET=0x380000
 CONFIG_SYS_SPI_U_BOOT_OFFS=0x100000
 CONFIG_DM_GPIO=y
 CONFIG_SPL_DM_SPI=y
@@ -79,7 +80,12 @@ CONFIG_OF_CONTROL=y
 CONFIG_SPL_OF_CONTROL=y
 CONFIG_SPL_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT_NO_COMPRESSION=y
+CONFIG_ENV_IS_IN_FAT=y
+CONFIG_ENV_IS_IN_MMC=y
+CONFIG_ENV_FAT_DEVICE_AND_PART=":1"
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
+CONFIG_SYS_MMC_ENV_PART=1
+CONFIG_SPL_ENV_IS_NOWHERE=y
 CONFIG_DM=y
 CONFIG_SPL_DM=y
 CONFIG_SPL_DM_SEQ_ALIAS=y
-- 
2.35.3

